File: animals/src/main/java/agh/ics/oop/model/Vector2d.java
package agh.ics.oop.model;
import java.util.Objects;

public class Vector2d {
    private final int x;
    private final int y;
    public Vector2d(int x,int y){
        this.x=x;
        this.y=y;
    }
    public int getX() {
        return x;
    }
    public int getY() {
        return y;
    }
    @Override
    public String toString(){
        return "(%d,%d)".formatted(this.x,this.y);
    }
    public boolean precedes(Vector2d other){
        return other.x>=this.x && other.y>=this.y;
    }
    public boolean follows(Vector2d other){
        return other.x<=this.x && other.y<=this.y;
    }
    public Vector2d add(Vector2d other){
        return new Vector2d(this.x+other.x,this.y+other.y);
    }
    public Vector2d subtract(Vector2d other){
        return new Vector2d(this.x-other.x,this.y-other.y);
    }
    public Vector2d upperRight(Vector2d other){ return new Vector2d(Math.max(other.x, this.x),Math.max(other.y, this.y));}
    public Vector2d lowerLeft(Vector2d other){ return new Vector2d(Math.min(other.x, this.x),Math.min(other.y, this.y));}
    public Vector2d opposite(){
        return new Vector2d(-this.x,-this.y);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Vector2d vector2d = (Vector2d) o;
        return this.x == vector2d.x && this.y == vector2d.y;
    }
    @Override
    public int hashCode() {
        return Objects.hash(this.x, this.y);
    }
}

File: animals_but_it_is_kotlin/src/main/kotlin/agh/ics/oop/model/Vector2d.kt
package agh.ics.oop.model

data class Vector2d(val x: Int, val y: Int) {

    operator fun compareTo(other: Vector2d): Int {
        if (other.x <= x && other.y < y || other.x < x && other.y <= y)
            return 1

        else if (this == other)
            return 0

        return -1
    }

    operator fun plus(other: Vector2d): Vector2d = Vector2d(x + other.x, y + other.y)

    operator fun minus(other: Vector2d): Vector2d = Vector2d(x - other.x, y - other.y)

    fun upperRight(other: Vector2d) = Vector2d(Math.max(other.x, x), Math.max(other.y, y))

    fun lowerLeft(other: Vector2d) = Vector2d(Math.min(other.x, x), Math.min(other.y, y))

    fun opposite() = Vector2d(-x, -y)

    override fun toString(): String = "($x,$y)"
}

fun MapDirection.toUnitVector(): Vector2d {
    return when (this) {
        MapDirection.NORTH -> Vector2d(0, 1)
        MapDirection.SOUTH -> Vector2d(0, -1)
        MapDirection.WEST -> Vector2d(-1, 0)
        MapDirection.EAST -> Vector2d(1, 0)
    }
}
